### PR TITLE
Add rmw_publisher_wait_for_all_acked support

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Related to ros2/rmw#295

### CODE REVIEW

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 1]
```diff
@@ -922,6 +924,31 @@ RMW_Connext_Publisher::assert_liveliness()
...
+    timeout.sec = static_cast<DDS_Long>(clamped_time.sec);
+    timeout.nanosec = static_cast<DDS_UnsignedLong>(clamped_time.nsec);
+  }
+
+  switch (DDS_DataWriter_wait_for_acknowledgments(this->dds_writer, &timeout)) {
+    case DDS_RETCODE_OK:
+      return RMW_RET_OK;
+    case DDS_RETCODE_TIMEOUT:
+      return RMW_RET_TIMEOUT;
+    default:
```
Comment (asorbini) [inline rmw_connextdds_common/src/common/rmw_impl.cpp:947]: Add a logging statement which also sets an error string for the rcl layer:

```suggestion
    default:
      {
        RMW_CONNEXT_LOG_ERROR_A_SET(
          "failed to wait for reader acknowledgements: dds_rc=%d", dds_rc)
        return RMW_RET_ERROR;
      }
```

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 2]
```diff
@@ -922,6 +924,31 @@ RMW_Connext_Publisher::assert_liveliness()
...
+  // replace below codes.
+  if (rmw_time_equal(wait_timeout, RMW_DURATION_INFINITE)) {
+    timeout = DDS_DURATION_INFINITE;
+  } else {
+    rmw_time_t clamped_time = rmw_dds_common::clamp_rmw_time_to_dds_time(wait_timeout);
+    timeout.sec = static_cast<DDS_Long>(clamped_time.sec);
+    timeout.nanosec = static_cast<DDS_UnsignedLong>(clamped_time.nsec);
+  }
+
+  switch (DDS_DataWriter_wait_for_acknowledgments(this->dds_writer, &timeout)) {
```
Comment (asorbini) [inline rmw_connextdds_common/src/common/rmw_impl.cpp:942]: Also, as a "coding style", I would prefer to cache the return code to a local variable, so we can also log it for the user (see below).
```suggestion
  const DDS_ReturnCode_t dds_rc =
    DDS_DataWriter_wait_for_acknowledgments(this->dds_writer, &timeout);
  switch (dds_rc) {
```

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 3]
```diff
@@ -922,6 +924,31 @@ RMW_Connext_Publisher::assert_liveliness()
   return RMW_RET_OK;
 }
 
+rmw_ret_t
+RMW_Connext_Publisher::wait_for_all_acked(rmw_time_t wait_timeout)
+{
+  DDS_Duration_t timeout;
```
Comment (asorbini) [inline rmw_connextdds_common/src/common/rmw_impl.cpp:930]: I would assume infinite timeout by default, and then change the `if()` below to check for `wait_timeout != RMW_DURATION_INFINITE`

```suggestion
  DDS_Duration_t timeout = DDS_DURATION_INFINITE;
```

**Review by asorbini — CHANGES_REQUESTED** [review]
Hi @Barry-Xu-2018, thank you for your submission!

I added a few comments and suggestions to make sure the code builds with RTI Connext DDS Micro, and a couple of other minor fixes.

Do you know if [rmw#295](https://github.com/ros2/rmw/issues/295) will be merged for Galactic?

**Comment by fujitatomoya** [comment]
> Do you know if rmw#295 will be merged for Galactic?

hopefully 🤞 

@wjwwood @clalancette @ivanpauno what do you think?

i guess we could have this in `rmw` for Galactic, then we can work on rcl, rclcpp, rosbag2 and tests accordingly for Turtle-H. but we do not have strong opinion about that.

**Review by fujitatomoya — COMMENTED** [review]
looks good to me with @asorbini 's comments

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 4]
```diff
@@ -922,6 +924,31 @@ RMW_Connext_Publisher::assert_liveliness()
...
+
+  // TODO(Barry): While rmw_connextdds_duration_from_ros_time() is changed to the public function,
+  // replace below codes.
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rmw_connextdds_common/src/common/rmw_impl.cpp:933]: Okay. I will move rmw_connextdds_duration_from_ros_time() and update codes.

File: rmw_connextdds_common/src/common/rmw_impl.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rmw_connextdds_common/src/common/rmw_impl.cpp:942]: Thanks for your suggestion. I update codes.

**Comment by wjwwood** [comment]
> hopefully 🤞

It's on my list to review/run CI for tonight.

Make sure it is rebased if you can, I've been having that stall the others.

**Review by asorbini — APPROVED** [review]
LGTM. Thank you for the updates @Barry-Xu-2018.

We should run some CI, but I assume that will happen to validate the rest of the changes in the other PRs.

**Comment by asorbini** [comment]
@fujitatomoya @Barry-Xu-2018 any updates on whether this feature will make it into Galactic?

Paging @clalancette too.

**Comment by Barry-Xu-2018** [comment]
@asorbini 

Galactic API/Feature Freeze.  
So this feature will be added to Turtle-H.

**Comment by Barry-Xu-2018** [comment]
Do rebase.

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 5]
```diff
@@ -920,6 +934,32 @@ RMW_Connext_Publisher::assert_liveliness()
...
 }
 
+rmw_ret_t
+RMW_Connext_Publisher::wait_for_all_acked(rmw_time_t wait_timeout)
+{
+  DDS_Duration_t timeout = DDS_DURATION_INFINITE;
+
+  if (!rmw_time_equal(wait_timeout, RMW_DURATION_INFINITE)) {
+    rmw_connextdds_duration_from_ros_time(&timeout, &wait_timeout);
+  }
```
Comment (ivanpauno) [inline rmw_connextdds_common/src/common/rmw_impl.cpp:944]: shouldn't the rmw_connextdds_duration_from_ros_time() function handle the infinity case already?

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 6]
```diff
@@ -219,6 +221,18 @@ rmw_connextdds_find_string_in_list(
   return false;
 }
 
+rmw_ret_t
+rmw_connextdds_duration_from_ros_time(
+  DDS_Duration_t * const duration,
+  const rmw_time_t * const ros_time)
```
Comment (ivanpauno) [inline rmw_connextdds_common/src/common/rmw_impl.cpp:227]: IMO, there's no reason to have a return code here, this function should never fail.

```suggestion
DDS_Duration_t
rmw_connextdds_duration_from_ros_time(
  const rmw_time_t * const ros_time)
```

File: rmw_connextddsmicro/src/rmw_api_impl_rtime.cpp [hunk 7]
```diff
@@ -414,6 +414,17 @@ rmw_publisher_assert_liveliness(
...
+  const rmw_publisher_t * publisher,
+  rmw_time_t wait_timeout
+)
+{
+  return rmw_api_connextdds_publisher_wait_for_all_acked(
+    publisher, wait_timeout);
+}
+
+
+mw_ret_t
```
Comment (ivanpauno) [inline rmw_connextddsmicro/src/rmw_api_impl_rtime.cpp:427]: :joy:

```suggestion
rmw_ret_t
```

File: rmw_connextdds_common/src/common/rmw_impl.cpp (hunk 5 above)
Comment (Barry-Xu-2018) [reply rmw_connextdds_common/src/common/rmw_impl.cpp:944]: The definition `DDS_Duration_t` is http://community.rti.com/rti-doc/510/ndds/doc/html/api_cpp/structDDS__Duration__t.html.
DDS_Long and DDS_UnsignedLong are 32-bit integer value.   
I think `rmw_connextdds_duration_from_ros_time()` cannot  covert infinite duration. So if the duration is infinite, directly use `DDS_DURATION_INFINITE`.

File: rmw_connextddsmicro/src/rmw_api_impl_rtime.cpp (hunk 7 above)
Comment (Barry-Xu-2018) [reply rmw_connextddsmicro/src/rmw_api_impl_rtime.cpp:427]: Sorry. My fault.

**Comment by Barry-Xu-2018** [comment]
Do rebase.

File: rmw_connextdds_common/src/common/rmw_impl.cpp (hunk 5 above)
Comment (Barry-Xu-2018) [reply rmw_connextdds_common/src/common/rmw_impl.cpp]: @ivanpauno  friendly ping.

File: rmw_connextddsmicro/src/rmw_api_impl_rtime.cpp (hunk 7 above)
Comment (ivanpauno) [reply rmw_connextddsmicro/src/rmw_api_impl_rtime.cpp]: Thanks for fixing it!

File: rmw_connextdds_common/src/common/rmw_impl.cpp (hunk 5 above)
Comment (ivanpauno) [reply rmw_connextdds_common/src/common/rmw_impl.cpp]: > I think rmw_connextdds_duration_from_ros_time() cannot correctly covert infinite duration (64-bit to 32-bit). So if the duration is infinite, directly use DDS_DURATION_INFINITE.

Sure, I mean that it should be handled inside `rmw_connextdds_duration_from_ros_time()` and not here.

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 8]
```diff
@@ -221,6 +221,19 @@ rmw_connextdds_find_string_in_list(
   return false;
 }
 
+DDS_Duration_t
+rmw_connextdds_duration_from_ros_time(
+  DDS_Duration_t * const duration,
+  const rmw_time_t * const ros_time)
```
Comment (ivanpauno) [inline rmw_connextdds_common/src/common/rmw_impl.cpp:227]: This is not currently matching the declaration in the header:

```suggestion
DDS_Duration_t
rmw_connextdds_duration_from_ros_time(
  const rmw_time_t * const ros_time)
```

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 9]
```diff
@@ -221,6 +221,19 @@ rmw_connextdds_find_string_in_list(
...
+{
+  rmw_time_t in_time = rmw_dds_common::clamp_rmw_time_to_dds_time(*ros_time);
+
+  DDS_Duration_t duration;
+  duration.sec = static_cast<DDS_Long>(in_time.sec);
+  duration.nanosec = static_cast<DDS_UnsignedLong>(in_time.nsec);
+  return duration;
+}
```
Comment (ivanpauno) [inline rmw_connextdds_common/src/common/rmw_impl.cpp:235]:
```suggestion
{
  if (rmw_time_equal(*ros_time, RMW_DURATION_INFINITE)) {
    return DDS_DURATION_INFINITE;
  }
  rmw_time_t in_time = rmw_dds_common::clamp_rmw_time_to_dds_time(*ros_time);

  DDS_Duration_t duration;
  duration.sec = static_cast<DDS_Long>(in_time.sec);
  duration.nanosec = static_cast<DDS_UnsignedLong>(in_time.nsec);
  return duration;
}
```

File: rmw_connextdds_common/src/common/rmw_impl.cpp (hunk 8 above)
Comment (ivanpauno) [reply rmw_connextdds_common/src/common/rmw_impl.cpp:227]: you can also use `const rmw_time_t * ros_time`, as the second `const` isn't really useful

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 10]
```diff
@@ -927,6 +940,32 @@ RMW_Connext_Publisher::assert_liveliness()
...
 }
 
+rmw_ret_t
+RMW_Connext_Publisher::wait_for_all_acked(rmw_time_t wait_timeout)
+{
+  DDS_Duration_t timeout = DDS_DURATION_INFINITE;
+
+  if (!rmw_time_equal(wait_timeout, RMW_DURATION_INFINITE)) {
+    timeout = rmw_connextdds_duration_from_ros_time(&wait_timeout);
+  }
```
Comment (ivanpauno) [inline rmw_connextdds_common/src/common/rmw_impl.cpp:950]:
```suggestion
  timeout = rmw_connextdds_duration_from_ros_time(&wait_timeout);
```

File: rmw_connextdds_common/src/common/rmw_impl.cpp (hunk 5 above)
Comment (ivanpauno) [reply rmw_connextdds_common/src/common/rmw_impl.cpp]: See https://github.com/ros2/rmw_connextdds/pull/20#discussion_r642479233 and https://github.com/ros2/rmw_connextdds/pull/20#discussion_r642480389.

Comment (Barry-Xu-2018) [reply rmw_connextdds_common/src/common/rmw_impl.cpp]: Get it. Thanks.

File: rmw_connextdds_common/src/common/rmw_impl.cpp (hunk 8 above)
Comment (Barry-Xu-2018) [reply rmw_connextdds_common/src/common/rmw_impl.cpp:227]: Yes.

File: rmw_connextdds_common/src/common/rmw_impl.cpp (hunk 9 above)
Comment (Barry-Xu-2018) [reply rmw_connextdds_common/src/common/rmw_impl.cpp:235]: thanks

File: rmw_connextdds_common/src/common/rmw_impl.cpp (hunk 8 above)
Comment (Barry-Xu-2018) [reply rmw_connextdds_common/src/common/rmw_impl.cpp:227]: Yes.

**Comment by Barry-Xu-2018** [comment]
@ivanpauno  Address your comments.

**Review by ivanpauno — APPROVED** [review]

**Pull request merged**